### DISCUSSION TITLE
Design and develop  ISO 14229-1 Road vehicles — Unified diagnostic services (UDS)  Application layer into project

### DISCUSSION THREAD

**Discussion by ajayjadhav900** [langroodi/Adaptive-AUTOSAR#39] [General]
I have ISO 14229-1 Road vehicles — Unified diagnostic services (UDS)  Application layer document.
how can I design and develop (vehicle diagnostic communication framework) application with UDS services  in your project.
Unified Diagnostic Services (UDS) is a diagnostic [communication protocol] https://en.wikipedia.org/wiki/Unified_Diagnostic_Services
with services like Diagnostic Session Control	, ECU RESET and few others.

<img width="576" alt="applayer" src="https://github.com/user-attachments/assets/269dd854-4313-4463-9ebc-82503d9db965">

**Comment by langroodi** [comment]
UDS is partially supported in this project. For instance, [this is the definition](https://github.com/langroodi/Adaptive-AUTOSAR/blob/master/src/ara/diag/ecu_reset_request.h) of the ECU reset.

Because the Adaptive AUTOSAR does not specify the implementation, then the UDS implementation can be different from a project to another. Here in this project, a routing mechanism is hire.

If you want to introduce a new UDS action, you need to inherit from the [UdsServiceRouter](https://github.com/langroodi/Adaptive-AUTOSAR/blob/master/src/ara/diag/routing/routable_uds_service.h) abstract class. The you can use the [UdsServiceRouter](https://github.com/langroodi/Adaptive-AUTOSAR/blob/master/src/ara/diag/routing/uds_service_router.h), to add/remove your new service.
